components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthnService.java

```diff
@@ -258,4 +302,110 @@ private boolean canAuthenticate(OAuthClientAuthenticator oAuthClientAuthenticato
 
         return oAuthClientAuthenticator.canAuthenticate(request, bodyContentMap, oAuthClientAuthnContext);
     }
+
+    /**
+     * Obtain the client authentication method configured for the application.
+     *
+     * @param clientId     Client ID of the application.
+     * @return Configured client authentication method for the application.
+     * @throws OAuthClientAuthnException OAuth Client Authentication Exception.
+     */
+    private List<String> getConfiguredClientAuthMethods(String clientId) throws OAuthClientAuthnException {
+
+        String tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+        try {
+            OAuthAppDO oAuthAppDO = OAuth2Util.getAppInformationByClientId(clientId, tenantDomain);
+            String tokenEndpointAuthMethod = oAuthAppDO.getTokenEndpointAuthMethod();
+            if (StringUtils.isNotBlank(tokenEndpointAuthMethod)) {
+                return Arrays.asList(tokenEndpointAuthMethod);
+            }
+        } catch (IdentityOAuth2Exception | InvalidOAuthClientException e) {
+            throw new OAuthClientAuthnException("Error occurred while retrieving app information for client id: " +
+                    clientId + " of tenantDomain: " + tenantDomain, OAuth2ErrorCodes.INVALID_REQUEST, e);
+        }
+        return Collections.emptyList();
+    }
+
+    /**
+     * Obtain the client ID of the application from the request.
+     *
+     * @param request                   Http servlet request.
+     * @param bodyContentMap            Content of the body of the request as a parameter map.
+     * @return Client ID of the application.
+     * @throws OAuthClientAuthnException OAuth Client Authentication Exception.
+     */
+    private String extractClientId(HttpServletRequest request, Map<String, List> bodyContentMap)
+            throws OAuthClientAuthnException {
+
+        String clientId = null;
+        for (OAuthClientAuthenticator oAuthClientAuthenticator : this.getClientAuthenticators()) {
+            try {
+                /* A new OAuthClientAuthnContext is used each time as we just need to extract the Client ID and we do
+                not want to overwrite the original OAuthClientAuthnContext.  */
+                OAuthClientAuthnContext oAuthClientAuthnContext = new OAuthClientAuthnContext();
+                clientId = oAuthClientAuthenticator.getClientId(request, bodyContentMap, oAuthClientAuthnContext);
+                if (StringUtils.isNotBlank(clientId)) {
+                    break;
+                }
+            } catch (OAuthClientAuthnException e) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Client ID cannot be extracted using the " + oAuthClientAuthenticator.getName());
```

Comment:
Log the exception as well.

Reply:
Fixed in 93533cd3fb2063210e5a4547007e6c93f4fc7479